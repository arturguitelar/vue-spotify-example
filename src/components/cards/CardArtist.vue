<template>
  <div v-if="artist" class="card card-artist">
    <h3 class="name">{{ artist.name }}</h3>
    <img :src="artist.imageUrl || 'https://via.placeholder.com/500'" />
    <div class="info">
      <p class="genres">{{ getGenres() }}</p>
      <p class="popularity">
        Popularidade: {{ artist.popularity }}
        <span :class="getBadge()">{{ getBadge() }}</span>
      </p>
    </div>
  </div>
  <div v-else>
    <p>Dados inv√°lidos</p>
  </div>
</template>

<script>
export default {
  name: "CardArtist",
  props: {
    artist: Object
  },
  methods: {
    getGenres() {
      if (this.artist.genres.length === 0) {
        return "sem categoria";
      }
      return this.artist.genres.join(" - ");
    },
    getBadge() {
      const popularity = this.artist.popularity;

      if (popularity >= 80) {
        return "hot";
      } else if (popularity >= 60) {
        return "cool";
      } else if (popularity >= 30) {
        return "ok";
      }

      return "underground";
    }
  }
};
</script>

<style>
.card-artist {
  display: flex;
  flex-direction: column;
}

.card-artist .name {
  position: absolute;
  background-color: var(--secondary-color);
  padding: 10px;
  opacity: 0.8;
  margin-top: 10px;
  font-style: italic;
  font-variant-caps: all-small-caps;
}

.card-artist img {
  width: 100%;
  height: auto;
}

.card-artist .info {
  display: flex;
  flex-direction: column;
  flex: 1;
  justify-content: space-between;
  padding: 10px;
  line-height: 1.2rem;
}

.card-artist .genres {
  color: var(--gray-text);
  font-style: italic;
  font-size: 12px;
  margin-bottom: 10px;
}

.card-artist .popularity {
  display: flex;
  justify-content: space-between;
  border-top: var(--border-default);
  padding-top: 10px;
}

.card-artist .popularity span {
  font-variant: small-caps;
  padding: 6px;
  color: white;
}

.popularity .hot {
  font-size: 30px;
  background: rgb(170, 14, 14);
  font-weight: bold;
  box-shadow: 3px 3px 3px 0px rgb(0 0 0 / 30%);
}

.popularity .cool {
  font-size: 26px;
  background: rgb(23, 134, 33);
  font-weight: bold;
  box-shadow: 3px 3px 3px 0px rgb(0 0 0 / 30%);
}

.popularity .ok {
  font-size: 20px;
  font-weight: bold;
  box-shadow: 3px 3px 3px 0px rgb(0 0 0 / 30%);
}

.popularity .underground {
  font-size: 14px;
  color: gray;
}
</style>
