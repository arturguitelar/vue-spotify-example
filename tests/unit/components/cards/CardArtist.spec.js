import { shallowMount } from "@vue/test-utils";
import CardArtist from "@/components/cards/CardArtist.vue";

const DATASET_CARD = "[data-card='artist']";
const DATASET_MESSAGE = "[data-card-message='empty']";
const IMAGE_URL = "https://via.placeholder.com/500";
const messages = {
  INVALID_DATA: "Dados invÃ¡lidos",
  NO_CATEGORY: "sem categoria"
};

describe("#CardArtist", () => {
  describe("#FAIL", () => {
    test("render invalid data if doesn't receive an artist object", () => {
      const wrapper = shallowMount(CardArtist);

      expect(wrapper.find(DATASET_CARD).exists()).toBe(false);
      expect(wrapper.find(DATASET_MESSAGE).exists()).toBe(true);
      expect(wrapper.text()).toBe(messages.INVALID_DATA);
    });

    test("render message if artist has no genres", () => {
      const wrapper = shallowMount(CardArtist, {
        props: {
          artist: {
            name: "Some Artist",
            imageUrl: IMAGE_URL,
            genres: [],
            popularity: 50
          }
        }
      });

      expect(wrapper.vm.getGenres()).toStrictEqual(messages.NO_CATEGORY);
      expect(wrapper.html()).toContain(messages.NO_CATEGORY);
    });
  });

  describe("#SUCCESS", () => {
    const defaultData = {
      name: "Some Artist",
      imageUrl: IMAGE_URL,
      genres: ["Some", "Genres", "Here"],
      popularity: 50
    };

    let wrapper;

    beforeEach(() => {
      wrapper = shallowMount(CardArtist, {
        props: { artist: defaultData }
      });
    });

    afterEach(() => {
      wrapper.unmount();
    });

    test("create a card with correct data", () => {
      expect(wrapper.vm.artist).toStrictEqual(defaultData);
      expect(wrapper.props()).toEqual({ artist: defaultData });
      expect(wrapper.find(DATASET_CARD).exists()).toBe(true);
      expect(wrapper.find(DATASET_MESSAGE).exists()).toBe(false);
    });

    test("get a string of genres to render in genres", () => {
      const expectedFormat = "Some - Genres - Here";

      expect(wrapper.vm.getGenres()).toStrictEqual(expectedFormat);
      expect(wrapper.text()).toContain(expectedFormat);
    });

    test("image should be placed by placeholder if no image url", () => {
      defaultData.imageUrl = "";

      wrapper = shallowMount(CardArtist, {
        props: { artist: defaultData }
      });

      expect(wrapper.find("img").attributes()).toEqual({
        src: IMAGE_URL
      });
    });

    test("badge score should be 'hot' if artist popularity >= 80", () => {
      scoreHelper(80, "hot");
    });

    test("badge score should be 'cool' if artist popularity >= 60", () => {
      scoreHelper(60, "cool");
    });

    test("badge score should be 'ok' if artist popularity >= 30", () => {
      scoreHelper(30, "ok");
    });

    test("badge score should be 'underground' if artist popularity < 30", () => {
      scoreHelper(29, "underground");
    });

    function scoreHelper(popularity, badge) {
      defaultData.popularity = popularity;

      wrapper = shallowMount(CardArtist, {
        props: { artist: defaultData }
      });

      expect(wrapper.find("span").classes()).toContain(badge);
      expect(wrapper.find("span").text()).toBe(badge);
    }

    describe("Snapshots", () => {
      test("should render a card correctly", () => {
        expect(wrapper.wrapperElement).toMatchSnapshot();
      });
    });
  });
});
