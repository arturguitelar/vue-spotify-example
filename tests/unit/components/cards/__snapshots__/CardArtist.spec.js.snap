// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#CardArtist #SUCCESS Snapshots should render a card correctly 1`] = `
<div
  class="card card-artist"
  data-card="artist"
>
  <h3
    class="name"
  >
    Some Artist
  </h3>
  <img
    src="https://via.placeholder.com/500"
  />
  <div
    class="info"
  >
    <p
      class="genres"
    >
      Some - Genres - Here
    </p>
    <p
      class="popularity"
    >
       Popularidade: 29 
      <span
        class="underground"
      >
        underground
      </span>
    </p>
  </div>
</div>
`;
